GoodreadsEDA/all_bks.py

```diff
@@ -21,52 +21,35 @@ def get_mv_release(bk_m):
                                        WHERE book_title="{bk_m}"
                                        ;""", conn)
         return df_rel["release_date"][0]
-    except IndexError as e:
-        print(f"{e} \n")
+    except IndexError:
+        traceback.print_exc()
         print("""Hmm... it looks as though the program cannot find the
 Release_Date table... ... did you run imdb_release.py yet?""")
 
 
 def create_even_samples(df, df2):
-    """Ballances the data so that the before and after values have an equal #
-    of sample sies."""
+    """Balances the data so that the before and after values have an equal #
+    of sample sites."""
 
     sample_diff = df.shape[0] - df2.shape[0]
     if sample_diff > 0:
-        df = df.sample(df2.shape[0])
-        return (df, df2)
+        return df.sample(df2.shape[0]), df2
     elif sample_diff < 0:
-        df2 = df2.sample(df.shape[0])
-        return (df, df2)
-    else:
-        return (df, df2)
+        return df, df2.sample(df.shape[0])
+    return df, df2
 
 
 # Create empty df to append each book to after preprocessing.
 df_all = pd.DataFrame()
 
 # Go through every title in the list above.
-for i, titl in enumerate(movie_titl):
-    bk = movie_titl[i]
+for title in movie_titles:
     # Get movie release date for movie and df of reviews for the book.
-    conn = sqlite3.connect(f"{os.getcwd()}/review_dbs/reviews.db")
-    mv_release = get_mv_release(bk)
-    # These two cases seemed to be giving me trouble, so I'm just hard coding them
-    # For now.
-    if titl == "Ender’s Game":
-        df = pd.read_sql_query(f"""SELECT * FROM Reviews
-                                   WHERE book_title = "Ender's Game (Ender's Saga, #1)"
-                                   ;""", conn)
-    elif titl == "Billy Lynn’s Long Halftime Walk":
-        df = pd.read_sql_query(f"""SELECT * FROM Reviews
-                               WHERE book_title LIKE "Billy Lynn%"
-                               ;""", conn)
-    else:
-        df = pd.read_sql_query(f"""SELECT * FROM Reviews WHERE book_title LIKE "{bk}%";""", conn)
-
-    conn.close()
+    with sqlite3.connect(f"{os.getcwd()}/review_dbs/reviews.db") as conn:
+        df = pd.read_sql_query("SELECT * FROM Reviews WHERE book_title LIKE ?;", conn, params=[title + '%'])
```

Comment:
🎉 SQL Statement parameterization 🎉 

It's more complex here than in a straight `cursor.execute` call, but the idea is that you don't really have to worry about escaping / formatting according to column types, and is generally the recommended method of executing SQL statements from code.

`params` takes a list of items, so in this case, it's just a list of one.

Reply:
I love `with`, this is way easier to read and for me to wrap my head around. I get in the rut of learning one way and forgetting to attempt better ways of handling things. I will pop through the rest of the files and make sure all SQL calls are formatted like this one. 

And, yeah, you're absolutely right. Doing the preferred method should save a lot of headaches with edge cases, such as the "Ender's Game" one you later noted.